src/python/pants/engine/engine.py

```diff
@@ -109,10 +112,18 @@ def product_request(self, product, subjects):
 
     # Throw handling.
     # TODO: See https://github.com/pantsbuild/pants/issues/3912
-    throw_roots = tuple(root for root, state in result_items if type(state) is Throw)
-    if throw_roots:
-      cumulative_trace = '\n'.join(self._scheduler.trace())
-      raise ExecutionError('Received unexpected Throw state(s):\n{}'.format(cumulative_trace))
+    throw_root_states = tuple(state for root, state in result_items if type(state) is Throw)
+    if throw_root_states:
+      if self._include_trace_on_error:
+        cumulative_trace = '\n'.join(self._scheduler.trace())
+        raise ExecutionError('Received unexpected Throw state(s):\n{}'.format(cumulative_trace))
+
+      if len(throw_root_states) == 1:
+        raise throw_root_states[0].exc
+      else:
+        raise ExecutionError('Multiple exceptions encountered:\n  {}'
```

Comment:
Hm. That's true.
A few options:
* dedup common errors
* for each root, display what we were going after and the associated error message. This might still be too much information for the end user case though.

I'm going to defer this to another round error message improvements if that's ok.